packages/cletus/src/agents/chat-agent.ts

```diff
@@ -1,211 +1,201 @@
+  if (!config) return;
+  
+  const types = config.getData().types;
+
+  // Get current DBA toolsets
+  const currentDBAToolsets = toolRegistry.getToolsets().filter(t => t.startsWith('dba:'));
+  const expectedDBAToolsets = types.map(t => getDBAToolsetName(t.name));
+
+  // Remove old DBA toolsets
+  for (const toolset of currentDBAToolsets) {
+    if (!expectedDBAToolsets.includes(toolset)) {
+      toolRegistry.unregisterToolset(toolset);
+    }
+  }
+
+  // Add/update DBA toolsets
+  for (const type of types) {
+    const toolsetName = getDBAToolsetName(type.name);
+    toolRegistry.unregisterToolset(toolsetName);
+    const dbaTools = createDBATools(type);
+    toolRegistry.registerToolset(toolsetName, dbaTools, getToolInstructions);
+  }
+
+  // Re-embed all DBA tools
+  for (const type of types) {
+    const toolsetName = getDBAToolsetName(type.name);
+    toolRegistry.reembedToolset(toolsetName);
+  }
+}
 
-- **dba:[type]**: Data operations for each custom data type defined in the system
-${dbaTools.map(tool => `  - ${Operations[tool.name].signature}`).join('\n')}
+/**
+ * Get tools based on current toolset or adaptive selection
+ */
+async function getActiveTools(ctx: CletusAIContext): Promise<RegisteredTool[]> {
+  const toolset = ctx.toolset;
+
+  if (toolset) {
+    // Use specific toolset
+    return toolRegistry.getToolset(toolset);
+  }
+
+  // Adaptive selection: use embeddings of recent user messages
+  const query = buildToolSelectionQuery(ctx.messages || []);
+  if (!query) {
+    // No user messages yet, return a default set of tools
+    return [
+      ...toolRegistry.getToolset('utility'),
+      ...toolRegistry.getToolset('planner').slice(0, 2),
+      ...toolRegistry.getToolset('clerk').slice(0, 3),
+    ];
+  }
+
+  // Select tools based on semantic similarity
+  // Always include utility tools
+  const utilityTools = toolRegistry.getToolset('utility');
+  const selectedTools = await toolRegistry.selectTools(
+    query,
+    ADAPTIVE_TOOLING.TOP_TOOLS_TO_SELECT - utilityTools.length,
+    ['utility'] // Exclude utility since we add it separately
+  );
+
+  return [...utilityTools, ...selectedTools];
+}
 
-Choose the appropriate agent based on what the user wants done.
-You need to provide a 'request' that includes all necessary details for the sub-agent to complete the task. This request should begin with human readable instructions followed by a technical description of what needs to be done (for example, a signature of one of the above tools with parameters filled in).
+/**
+ * Create the main chat agent with adaptive tooling
+ */
+export function createChatAgent(ai: CletusAI) {
+  // Create all toolsets
+  const toolsets = createToolsets(ai);
+  
+  // Initialize the tool registry
+  initializeToolRegistry(ai, toolsets);
+
+  // Register listener for type changes to update DBA tools dynamically
+  const config = ai.config.defaultContext?.config;
+  if (config) {
+    config.onTypeChange(() => {
```

Comment:
Let's register listers by a reference name so we don't have old chat instances receiving notifications